File: lg1801.cpp
#include <vector>
#include <cstdio>
#include <algorithm>
#include <queue>

using namespace std;

priority_queue<int, vector<int>, greater<int>> snd;
priority_queue<int> fst;
int sz = 1;
int a[200003], b[200003];

int main() {
    int n, m;
    scanf("%d %d", &n, &m);

    for (int i=0; i<n; ++i) scanf("%d", &a[i]);
    for (int i=0; i<m; ++i) scanf("%d", &b[i]);
    
    for (int j=0,i=0; j<m; ) {
        while (i<b[j]) {
            if (fst.empty() || a[i]<=fst.top()) 
                fst.push(a[i]);
            else      
                snd.push(a[i]);
            ++i;
        }

        while (fst.size()>sz) snd.push(fst.top()), fst.pop();
        while (fst.size()<sz) fst.push(snd.top()), snd.pop();

        printf("%d\n", fst.top());
        ++j;
        ++sz;
    }

    return 0;
}

File: lg1801.java
import java.util.*;

public class lg1801 {

    static int sz = 1;
    static Queue<Integer> fst = new PriorityQueue<Integer>(10, new Comparator<Integer>() {
        @Override
        public int compare(Integer a, Integer b) {
            return b.compareTo(a);
        }
    });

    static Queue<Integer> snd = new PriorityQueue<Integer>();

    public static void main(String[] args) {
        Scanner in = new Scanner(System.in);
        
        int n, m;
        n = in.nextInt(); m = in.nextInt();

        int[] a = new int[n];
        int[] b = new int[m];

        for (int i=0; i<n; ++i) a[i] = in.nextInt();
        for (int i=0; i<m; ++i) b[i] = in.nextInt();

        for (int j=0,i=0; j<m; ) {
            while (i<b[j]) {
                if (fst.isEmpty() || a[i]<=fst.peek()) 
                    fst.add(a[i]);
                else      
                    snd.add(a[i]);
                ++i;
            }

            while (fst.size()>sz) snd.add(fst.poll());
            while (fst.size()<sz) fst.add(snd.poll());

            System.out.println(fst.peek());
            ++j;
            ++sz;
        }
    }
}